bundles/org.openhab.binding.modbus/src/main/java/org/openhab/binding/modbus/internal/profiles/ModbusGainOffsetProfile.java

```diff
@@ -0,0 +1,248 @@
+ *
+ *
+ * @author Sami Salonen - Initial contribution
+ */
+@NonNullByDefault
+public class ModbusGainOffsetProfile implements StateProfile {
+
+    private final Logger logger = LoggerFactory.getLogger(ModbusGainOffsetProfile.class);
+    private static final String PREGAIN_OFFSET_PARAM = "pre-gain-offset";
+    private static final String GAIN_PARAM = "gain";
+
+    private final ProfileCallback callback;
+    private final ProfileContext context;
+
+    private Optional<QuantityType<?>> pregainOffset;
+    private Optional<QuantityType<?>> gain;
+
+    public ModbusGainOffsetProfile(ProfileCallback callback, ProfileContext context) {
+        this.callback = callback;
+        this.context = context;
+        {
+            Object rawOffsetValue = orDefault("0", this.context.getConfiguration().get(PREGAIN_OFFSET_PARAM));
+            logger.debug("Configuring profile with {} parameter '{}'", PREGAIN_OFFSET_PARAM, rawOffsetValue);
+            pregainOffset = parameterAsQuantityType(PREGAIN_OFFSET_PARAM, rawOffsetValue, Units.ONE);
+
+        }
+        {
+            Object gainValue = orDefault("1", this.context.getConfiguration().get(GAIN_PARAM));
+            logger.debug("Configuring profile with {} parameter '{}'", GAIN_PARAM, gainValue);
+            gain = parameterAsQuantityType(GAIN_PARAM, gainValue);
+
+        }
+    }
+
+    public boolean isValid() {
+        return pregainOffset.isPresent() && gain.isPresent();
+    }
+
+    public Optional<QuantityType<?>> getPregainOffset() {
+        return pregainOffset;
+    }
+
+    public Optional<QuantityType<?>> getGain() {
+        return gain;
+    }
+
+    @Override
+    public ProfileTypeUID getProfileTypeUID() {
+        return ModbusProfiles.GAIN_OFFSET;
+    }
+
+    @Override
+    public void onStateUpdateFromItem(State state) {
+        // no-op
+    }
+
+    @Override
+    public void onCommandFromItem(Command command) {
+        Type result = applyGainOffset(command, false);
+        if (result instanceof Command) {
+            logger.trace("Command '{}' from item, sending converted '{}' state towards handler.", command, result);
+            callback.handleCommand((Command) result);
+        }
+    }
+
+    @Override
+    public void onCommandFromHandler(Command command) {
+        Type result = applyGainOffset(command, true);
+        if (result instanceof Command) {
+            logger.trace("Command '{}' from handler, sending converted '{}' command towards item.", command, result);
+            callback.sendCommand((Command) result);
+        }
+    }
+
+    @Override
+    public void onStateUpdateFromHandler(State state) {
+        State result = (State) applyGainOffset(state, true);
+        logger.trace("State update '{}' from handler, sending converted '{}' state towards item.", state, result);
+        callback.sendUpdate(result);
+    }
+
+    @SuppressWarnings({ "unchecked", "rawtypes" })
```

Comment:
The missing types for the generics bother me a bit. Maybe the whole class would've been simpler if you did the calculation in the `BigDecimal` domain rather in `QuantityType`? I.e. convert the `QuantityType` to `BigDecimal` perform the transformation and convert it back to `QuantityType`?

Reply:
I'm not sure to be honest, there's quite a bit of logic with units (e.g. dividing quantitype by quantitype) and raising proper error messages etc.

In any case, I've added the more "boilerplate" to avoid supressing the warnings  https://github.com/openhab/openhab-addons/pull/9980/commits/e4c481c642bf4f10ff3b5ad11f9835c104113146.

Reply:
You've more insights of the code. Was more a gut feeling, that it could be simpler operating in the `BigDecimal` domain. The code looks better, now, IMHO.

Reply:
Thanks. Let's keep it like this...